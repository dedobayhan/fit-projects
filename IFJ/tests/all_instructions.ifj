var int gint,
	double gdbl,
	string gstr;

// Deklarace
void ret_void();
int ret_int(int);
double ret_dbl(double);
string ret_str(string);

//
// Funkce, ktere prijimaji a vraci dany "datovy typ"
//
void ret_void()
{
	return;
}

int ret_int(int i)
{
	return i;
}

double ret_dbl(double d)
{
	return d;
}

string ret_str(string s)
{
	return s;
}

// Deklarace - definice bude na konci
int dec(int);

// Pomocna funkce s vice parametry
int add(int a; int b; int c; int d; int e)
var int pom;
{
	pom := 0;
	pom := pom + a;
	pom := pom + b;
	pom := pom + c;
	pom := pom + d;
	pom := pom + e;

	return pom;
}

//
// Main
//
void main()
var int i,
	double d,
	string s;
{
	// Test vypisu neinicializovane hodnoty
	printstring("Test vypisu neinicializovane hodnoty: ");
	printint(i); printstring(" ");
	printdouble(d); printstring(" ");
	printstring(s); printstring("\n");

	// Goto
	printstring("Goto ");
	goto label;
	printstring("CHYBA\n");
	return;
	label:
	printstring("OK\n");

	// "Cyklus"
	i := 5;
	printstring("Cyklus ");
	printint(i);
	start:
	i := dec(i);
	printint(i);
	if i>0 goto start;
	printstring("\n");

	// If true
	printstring("If true: ");
	if 99 goto lab1;
	printstring("CHYBA\n");
	return;
	lab1:
	if 99.9 goto lab2;
	printstring("CHYBA\n");
	return;
	lab2:
	if "abcd" goto lab3;
	printstring("CHYBA\n");
	return;
	lab3:
	printstring("OK\n");

	// If true (vyraz)
	i := 10;
	d := 20.0;
	s := "abcd";
	printstring("If true (vyraz): ");
	if i < d goto lab11;
	printstring("CHYBA\n");
	return;
	lab11:
	if s<"abcdef" goto lab21;
	printstring("CHYBA\n");
	return;
	lab21:
	if i^d>=100 goto lab31;
	printstring("CHYBA\n");
	return;
	lab31:
	printstring("OK\n");

	// If false
	printstring("If false: ");
	if 0 goto lab4;
	printstring("OK1 ");
	lab4:
	if 0.0 goto lab5;
	printstring("OK2 ");
	lab5:
	if "" goto lab6;
	printstring("OK3\n");
	lab6:

	// If false (vyraz)
	i := 1000;
	d := 0.0;
	s := "";
	printstring("If false (vyraz): ");
	if 1000 == i+1 goto lab41;
	printstring("OK1 ");
	lab41:
	if 0 != 0.0 != d goto lab51;
	printstring("OK2 ");
	lab51:
	if (s+""+s+s) goto lab61;
	printstring("OK3\n");
	lab61:

	// Jednoduche volani funkci
	printstring("Volani funkci: ");
	ret_void();
	printint(ret_int(1)); printstring(" ");
	ret_int(1000);
	printdouble(ret_dbl(1.0)); printstring(" ");
	ret_dbl(100000000000.0);
	printstring(ret_str("abcd")); printstring("\n");
	ret_str("abcdefghijklmnopqrstuvwzyz");

	// Volani funkce z funkce z funkce ...
	printstring("Volani funkce z funkce z funkce: ");
	printstring(ret_str("a" + ret_str("b" + ret_str("c" + ret_str("d" +  ret_str("e" + ret_str("f")))))));
	printstring("\n");

	// Zavorky
	printstring("Zavorky: ");
	printint((((((((((((1))))))))))));
	printstring("\n");

	//
	// Vyrazy
	//

	// Scitani
	gint := 5;
	gdbl := 10.0;
	gstr := "123456789";
	i := 1;
	d := 30.0;
	s := "987654321";
	printstring("Scitani: ");
	printint(gint + i + 1); printstring(" ");
	printdouble(100.000 + gdbl + d); printstring(" ");
	printstring(gstr + "0" + s); printstring("\n");

	// Konkatenace
	printstring("Konkatenace: " + "" + "..... " + "OK" + "\n");

	// Odcitani
	gint := 50;
	gdbl := 100.0;
	i := 10;
	d := 300.1;
	printstring("Odcitani: ");
	printint(gint - i - 1); printstring(" ");
	printdouble(100.000 - gdbl - d + 0.1); printstring("\n");

	// Nasobeni
	gint := 5;
	gdbl := 10.0;
	i := 2;
	d := 3.0;
	printstring("Nasobeni: ");
	printint(gint * i * 2); printstring(" ");
	printdouble(100.000 * gdbl * d * 0.001); printstring("\n");

	// Deleni (divint)
	gint := 20;
	i := 2;
	printstring("Deleni (divint): ");
	printint(gint div i div 3); printstring("\n");

	// Deleni
	gint := 5;
	gdbl := 10.0;
	i := 2;
	d := 3.0;
	printstring("Deleni: ");
	printdouble(gint / i / 2); printstring(" ");
	printdouble(100.000 / gdbl / d / 0.001); printstring("\n");

	// Mocnina
	gint := 2;
	gdbl := 1.0;
	i := 3;
	d := 3.0;
	printstring("Mocnina: ");
	printdouble(gint ^ i ^ 2); printstring(" ");
	printdouble(2.0 ^ d ^ gdbl); printstring("\n");

	// Porovnani (int)
	gint := 1;
	i := 2;
	printstring("Porovnani (int): ");
	printint(gint < i < 3); printstring(" ");
	printint(i < gint < 1); printstring(" ");
	printint(gint > i > 3); printstring(" ");
	printint(i > gint > 1); printstring(" ");
	printint(gint == i); printstring(" ");
	printint(i != gint); printstring(" ");
	printint(gint <= i); printstring(" ");
	printint(i >= gint); printstring(" ");
	printint(3 < 2 < 1.0); printstring("\n");

	// Porovnani (double)
	gdbl := 1.0;
	d := 2.0;
	printstring("Porovnani (double): ");
	printint(gdbl < d < 3); printstring(" ");
	printint(d < gdbl < 1); printstring(" ");
	printint(gdbl > d > 3); printstring(" ");
	printint(d > gdbl > 1); printstring(" ");
	printint(gdbl == d); printstring(" ");
	printint(d != gdbl); printstring(" ");
	printint(gdbl <= d); printstring(" ");
	printint(d >= gdbl); printstring(" ");
	printint(3.0 < 2.0 < 1.0); printstring("\n");

	// Porovnani (string)
	gstr := "abc";
	s := "abcd";
	printstring("Porovnani (string): ");
	printint(gstr < s); printstring(" ");
	printint(s < gstr); printstring(" ");
	printint(gstr > s); printstring(" ");
	printint(s > gstr); printstring(" ");
	printint(gstr == "abc"); printstring(" ");
	printint(s != "abcd"); printstring(" ");
	printint(gstr <= "abc"); printstring(" ");
	printint(s >= "abcd"); printstring(" ");
	printint(s < gstr < 1.0); printstring("\n");

	// Priority
	printstring("Priority: ");
	printint(1+2*3+1); printstring(" ");
	printdouble(1+6/3+1); printstring(" ");
	printint(1+2 div 3+1); printstring(" ");
	printdouble(1+2^3+1); printstring("\n");

	// Slozeny vyraz (zavorky)
	printstring("Slozeny vyraz (zavorky): ");
	printint((1+2)*(3+1)); printstring(" ");
	printdouble((1-2)/(3-1)); printstring(" ");
	printint((1+2)div(3+1)); printstring(" ");
	printdouble((1+2)^(1+1)); printstring("\n");

	// Prirazeni
	printstring("Prirazeni: ");
	i := 1; printint(i); printstring(" ");
	i := 1 < 0; printint(i); printstring(" ");
	i := dec(1); printint(i); printstring(" ");
	i := dec(0); printint(i); printstring(" ");
	i := ((1+1+1+1)); printint(i); printstring(" ");
	printstring("\n");

	// Prazdny prikaz
	; ; ; ; ; ; label00: ; ; ; ; ;

	// Volani funkce s vice parametry
	printstring("Volani funkce s vice parametry: ");
	printint(add(1,2,3,add(1,2,3,4,5),add(1,2,3,4,5)));
	printstring("\n");
}

// Deklarace - definice bude na konci
int dec(int i)
{
	return i-2;
}

// Deklarace na konci programu
void ret_void();
int ret_int(int);
double ret_dbl(double);
string ret_str(string);
