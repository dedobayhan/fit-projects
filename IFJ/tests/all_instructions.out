Test vypisu neinicializovane hodnoty: 0 0 
Goto OK
Cyklus 531-1
If true: OK
If true (vyraz): OK
If false: OK1 OK2 OK3
If false (vyraz): OK1 OK2 OK3
Volani funkci: 1 1 abcd
Volani funkce z funkce z funkce: abcdef
Zavorky: 1
Scitani: 7 140 1234567890987654321
Konkatenace: ..... OK
Odcitani: 39 -300
Nasobeni: 20 3
Deleni (divint): 3
Deleni: 1.25 3333.33
Mocnina: 512 8
Porovnani (int): 1 1 0 0 0 1 1 1 1
Porovnani (double): 1 1 0 0 0 1 1 1 1
Porovnani (string): 1 0 0 1 1 0 1 1 1
Priority: 8 4 2 10
Slozeny vyraz (zavorky): 12 -0.5 0 9
Prirazeni: 1 0 -1 -2 4 
Volani funkce s vice parametry: 36
